arch/arm/core/cortex_m/pm_s2ram.S

```diff
@@ -14,58 +14,132 @@
 #include <zephyr/arch/cpu.h>
 #include <zephyr/arch/common/pm_s2ram.h>
 
+/**
+ * Macro expanding to the (integer literal) offset of field
+ * `sr_name` in the __cpu_context structure. This macro has
+ * to be implemented in C because GEN_OFFSET_SYM provides the
+ * offsets as C preprocessor definitions; they are not available
+ * by the time the assembler macros are processed.
+ *
+ * See also: `arch/arm/core/offsets/offsets_aarch32.c`
+ */
+#define CPU_CTX_SR_OFFSET(sr_name) \
+	___cpu_context_t_ ## sr_name ## _OFFSET
+
+/**
+ * Macros used to save / load a special register in __cpu_context.
+ * These also have to be implemented in C due to CPU_CTX_SR_OFFSET.
+ */
+#define SAVE_SPECIAL_REG(sr_name, cpu_ctx_reg, tmp_reg)	\
+	mrs	tmp_reg, sr_name;			\
+	str	tmp_reg, [cpu_ctx_reg, # CPU_CTX_SR_OFFSET(sr_name)];
+
+#define RESTORE_SPECIAL_REG(sr_name, cpu_ctx_reg, tmp_reg)		\
+	ldr	tmp_reg, [cpu_ctx_reg, # CPU_CTX_SR_OFFSET(sr_name)];	\
+	msr	sr_name, tmp_reg;
+
 _ASM_FILE_PROLOGUE
 
 GTEXT(pm_s2ram_mark_set)
 GTEXT(pm_s2ram_mark_check_and_clear)
 GDATA(_cpu_context)
 
+# Pushes registers r4~r12 and lr on the stack.
+# r0 is unmodified, other GPRs may be overwritten.
+.macro push_gprs
+#ifndef CONFIG_ARMV7_M_ARMV8_M_MAINLINE
+	/* ARMv6-M `push` only supports r0~r7 and LR */
+	push	{ r4-r7 }
+	mov	r1, r8
+	mov	r2, r9
+	mov	r3, r10
+	mov	r4, r11
+	mov	r5, r12
+	push	{ r1-r5, lr }
+#else
+	/* ARMv7-M and ARMv8-M have no limitation on `push` */
+	push	{r4-r12, lr}
+#endif /* !CONFIG_ARMV7_M_ARMV8_M_MAINLINE */
+.endm
+
+# Pops registers r4~r12 and lr from the stack
+# r0 is unmodified, other GPRs may be overwritten.
+.macro pop_gprs
+#ifndef CONFIG_ARMV7_M_ARMV8_M_MAINLINE
+	/* ARMv6-M `pop` only supports r0~r7 and PC */
+	pop	{ r1-r6 } /* can't pop lr to itself... */
+	mov 	lr, r6
+	mov 	r12, r5
+	mov 	r11, r4
+	mov 	r10, r3
+	mov 	r9, r2
+	mov 	r8, r1
+	pop	{ r4-r7 }
+#else
+	/* ARMv7-M and ARMv8-M have no limitation on `pop` */
+	pop	{r4-r12, lr}
+#endif /* !CONFIG_ARMV7_M_ARMV8_M_MAINLINE */
+.endm
+
+# Saves all special registers in the provided __cpu_context structure.
+.macro save_special_registers cpu_ctx, tmp_reg
+	SAVE_SPECIAL_REG(msp,		\cpu_ctx, \tmp_reg)
+	SAVE_SPECIAL_REG(psp,		\cpu_ctx, \tmp_reg)
+	SAVE_SPECIAL_REG(apsr,		\cpu_ctx, \tmp_reg)
+	SAVE_SPECIAL_REG(ipsr,		\cpu_ctx, \tmp_reg)
+	SAVE_SPECIAL_REG(epsr,		\cpu_ctx, \tmp_reg)
```

Comment:
EPSR reads as 0 and writes are ignored.
IPSR can be read but writes are ignored too.
See B1.4.2 of Armv6-m Architecture Reference Manual and B3.5 of the Armv8-m Architecture Reference Manual.

So trying to save these registers has no effect. These should be removed.